# Predict Loan Eligibility for Dream Housing Finance company

Dream Housing Finance company deals in all kinds of home loans. They have presence across all urban, semi urban and rural areas. Customer first applies for home loan and after that company validates the customer eligibility for loan.

Company wants to automate the loan eligibility process (real time) based on customer detail provided while filling online application form. These details are Gender, Marital Status, Education, Number of Dependents, Income, Loan Amount, Credit History and others. To automate this process, they have provided a dataset to identify the customers segments that are eligible for loan amount so that they can specifically target these customers. 



# Data Dictionary

### Train file: CSV containing the customers for whom loan eligibility is known as 'Loan_Status'

- Variable	                        -   Description
- Loan_ID                 : 	     Unique Loan ID
- Gender	              :          Male/ Female
- MarriedApplicant        :           married (Y/N)
- Dependents	          :          Number of dependents
- Education	Applicant     :       Education (Graduate/ Under Graduate)
- Self_Employed	          :       Self employed (Y/N)
- ApplicantIncome	      :       Applicant income
- CoapplicantIncome	      :      Coapplicant income
- LoanAmount	          :      Loan amount in thousands
- Loan_Amount_Term	      :     Term of loan in months
- Credit_History	      :     credit history meets guidelines
- Property_Area           :     Urban/ Semi Urban/ Rural
- Loan_Status	          :      (Target) Loan approved (Y/N)

### Test file: CSV containing the customer information for whom loan eligibility is to be predicted


# - Variable	                        -   Description
- Loan_ID                 : 	     Unique Loan ID
- Gender	              :          Male/ Female
- MarriedApplicant        :           married (Y/N)
- Dependents	          :          Number of dependents
- Education	Applicant     :       Education (Graduate/ Under Graduate)
- Self_Employed	          :       Self employed (Y/N)
- ApplicantIncome	      :       Applicant income
- CoapplicantIncome	      :      Coapplicant income
- LoanAmount	          :      Loan amount in thousands
- Loan_Amount_Term	      :     Term of loan in months
- Credit_History	      :     credit history meets guidelines
- Property_Area           :     Urban/ Semi Urban/ Rural



### 1. Loading data


In [1]:
import pandas as pd
import numpy as np

In [2]:
train=pd.read_csv("C:\\Users\\DELL\\Downloads\\Internship task\\loan data\\train_ctrUa4K.csv")
train.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

In [3]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            614 non-null    object 
 1   Gender             601 non-null    object 
 2   Married            611 non-null    object 
 3   Dependents         599 non-null    object 
 4   Education          614 non-null    object 
 5   Self_Employed      582 non-null    object 
 6   ApplicantIncome    614 non-null    int64  
 7   CoapplicantIncome  614 non-null    float64
 8   LoanAmount         592 non-null    float64
 9   Loan_Amount_Term   600 non-null    float64
 10  Credit_History     564 non-null    float64
 11  Property_Area      614 non-null    object 
 12  Loan_Status        614 non-null    object 
dtypes: float64(4), int64(1), object(8)
memory usage: 62.5+ KB


### 2.Data Preprocessing:

In [4]:
train.isnull().sum()

Loan_ID               0
Gender               13
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

In [5]:
# Train Categorical Variables Missisng values
train['Gender'].fillna(train['Gender'].mode()[0], inplace=True)
train ['Married'].fillna(train['Married'].mode()[0],inplace=True)
train['Dependents'].fillna(train['Dependents'].mode()[0], inplace=True)
train['Self_Employed'].fillna(train['Self_Employed'].mode()[0], inplace=True)
train['Credit_History'].fillna(train['Credit_History'].mode()[0], inplace=True)

# Train Numerical Variables Missing Values
train['Loan_Amount_Term'].fillna(train['Loan_Amount_Term'].mode()[0], inplace=True)
train['LoanAmount'].fillna(train['LoanAmount'].median(), inplace=True)


In [6]:
train.isnull().sum()

Loan_ID              0
Gender               0
Married              0
Dependents           0
Education            0
Self_Employed        0
ApplicantIncome      0
CoapplicantIncome    0
LoanAmount           0
Loan_Amount_Term     0
Credit_History       0
Property_Area        0
Loan_Status          0
dtype: int64

In [7]:
train.columns

Index(['Loan_ID', 'Gender', 'Married', 'Dependents', 'Education',
       'Self_Employed', 'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Property_Area', 'Loan_Status'],
      dtype='object')

In [8]:
train_copy=train.copy()

In [9]:
train_copy.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0       128.0             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

### 3.LabelEncoding:


In [10]:
from sklearn.preprocessing import LabelEncoder
l=["Gender","Married","Education","Self_Employed","Property_Area","Loan_Status"]

for i in l:
    le = LabelEncoder()
    le.fit(train[i])
    le.classes_
    
    train[i]=le.transform(train[i])
    

In [11]:
train.head()

Loan_ID  Gender  Married Dependents  Education  Self_Employed  \
0  LP001002       1        0          0          0              0   
1  LP001003       1        1          1          0              0   
2  LP001005       1        1          0          0              1   
3  LP001006       1        1          0          1              0   
4  LP001008       1        0          0          0              0   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0       128.0             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History  Property_Area  Loan_Status  
0             1.0              2            1  
1             1.0              0            0  
2             1.0              2            1  
3             1.0              2            1  
4             1.0              2            1

In [12]:
train[["Dependents"]]=train_copy["Dependents"].replace(to_replace={'3+':3})
train.head()

Loan_ID  Gender  Married Dependents  Education  Self_Employed  \
0  LP001002       1        0          0          0              0   
1  LP001003       1        1          1          0              0   
2  LP001005       1        1          0          0              1   
3  LP001006       1        1          0          1              0   
4  LP001008       1        0          0          0              0   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0       128.0             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History  Property_Area  Loan_Status  
0             1.0              2            1  
1             1.0              0            0  
2             1.0              2            1  
3             1.0              2            1  
4             1.0              2            1

In [13]:
import matplotlib.pyplot as plt
import seaborn as sns
plt.figure(figsize=(15,8))
sns.heatmap(train.corr(), cmap="rainbow", annot=True,linewidth=0.1)


### 4.Seprate dependent and independent variable:

In [14]:
# Separting the Variable into Independent and Dependent 
X = train.drop(["Loan_Status","Loan_ID"],axis=1)
y = train['Loan_Status']



In [15]:
X

Gender  Married Dependents  Education  Self_Employed  ApplicantIncome  \
0         1        0          0          0              0             5849   
1         1        1          1          0              0             4583   
2         1        1          0          0              1             3000   
3         1        1          0          1              0             2583   
4         1        0          0          0              0             6000   
..      ...      ...        ...        ...            ...              ...   
609       0        0          0          0              0             2900   
610       1        1          3          0              0             4106   
611       1        1          1          0              0             8072   
612       1        1          2          0              0             7583   
613       0        0          0          0              1             4583   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0                  0.0       128.0             360.0             1.0   
1               1508.0       128.0             360.0             1.0   
2                  0.0        66.0             360.0             1.0   
3               2358.0       120.0             360.0             1.0   
4                  0.0       141.0             360.0             1.0   
..                 ...         ...               ...             ...   
609                0.0        71.0             360.0             1.0   
610                0.0        40.0             180.0             1.0   
611              240.0       253.0             360.0             1.0   
612                0.0       187.0             360.0             1.0   
613                0.0       133.0             360.0             0.0   

     Property_Area  
0                2  
1                0  
2                2  
3                2  
4                2  
..             ...  
609              0  
610              0  
611              2  
612              2  
613              1  

[614 rows x 11 columns]

In [16]:
print(X.shape)
print(y.shape)

(614, 11)
(614,)


### 5.Train test split:

In [17]:
# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test,y_train,y_test = train_test_split(X, y,train_size=0.80, test_size = 0.20, random_state =0)

In [18]:
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)


(491, 11)
(491,)
(123, 11)
(123,)


### 6. Scaling:

In [19]:
# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

### 7.KNN

In [20]:
#Import knearest neighbors Classifier model
from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics

for i in range(0,10):
    i = i+1
    knn = KNeighborsClassifier(n_neighbors=i)
    #Train the model using the training sets
    knn.fit(X_train,y_train)
    #Predict the response for test dataset
    y_pred=knn.predict(X_test)
    
   
    #accuracy calculation
    acc=metrics.accuracy_score(y_test,y_pred)
    print("Accuracy of Knn for k=",i,'is:',acc)
    

Accuracy of Knn for k= 1 is: 0.7398373983739838
Accuracy of Knn for k= 2 is: 0.6504065040650406
Accuracy of Knn for k= 3 is: 0.7560975609756098
Accuracy of Knn for k= 4 is: 0.7154471544715447
Accuracy of Knn for k= 5 is: 0.7804878048780488
Accuracy of Knn for k= 6 is: 0.7886178861788617
Accuracy of Knn for k= 7 is: 0.8048780487804879
Accuracy of Knn for k= 8 is: 0.8130081300813008
Accuracy of Knn for k= 9 is: 0.8130081300813008
Accuracy of Knn for k= 10 is: 0.8211382113821138


From above here for K=10 which gives 82.11% of accuracy which is good.

In [21]:
# NOW WITH K=10
knn = KNeighborsClassifier(n_neighbors=10)
knn.fit(X_train,y_train)
pred = knn.predict(X_test)

print(metrics.confusion_matrix(y_test,pred))

print(metrics.classification_report(y_test,pred))

[[17 16]
 [ 6 84]]
             precision    recall  f1-score   support

          0       0.74      0.52      0.61        33
          1       0.84      0.93      0.88        90

avg / total       0.81      0.82      0.81       123



From above confusion matrix there are 101 samples are correctly classified and 22 samples are missclassified. The accuracy of KNN for k=10 is 82.11%.

### Prediction on test data using KNN


In [22]:
df=pd.read_csv("C:\\Users\\DELL\\Downloads\\Internship task\\loan data\\test_lAUu6dG.csv")
df

Loan_ID Gender Married Dependents     Education Self_Employed  \
0    LP001015   Male     Yes          0      Graduate            No   
1    LP001022   Male     Yes          1      Graduate            No   
2    LP001031   Male     Yes          2      Graduate            No   
3    LP001035   Male     Yes          2      Graduate            No   
4    LP001051   Male      No          0  Not Graduate            No   
..        ...    ...     ...        ...           ...           ...   
362  LP002971   Male     Yes         3+  Not Graduate           Yes   
363  LP002975   Male     Yes          0      Graduate            No   
364  LP002980   Male      No          0      Graduate            No   
365  LP002986   Male     Yes          0      Graduate            No   
366  LP002989   Male      No          0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5720                  0       110.0             360.0   
1               3076               1500       126.0             360.0   
2               5000               1800       208.0             360.0   
3               2340               2546       100.0             360.0   
4               3276                  0        78.0             360.0   
..               ...                ...         ...               ...   
362             4009               1777       113.0             360.0   
363             4158                709       115.0             360.0   
364             3250               1993       126.0             360.0   
365             5000               2393       158.0             360.0   
366             9200                  0        98.0             180.0   

     Credit_History Property_Area  
0               1.0         Urban  
1               1.0         Urban  
2               1.0         Urban  
3               NaN         Urban  
4               1.0         Urban  
..              ...           ...  
362             1.0         Urban  
363             1.0         Urban  
364             NaN     Semiurban  
365             1.0         Rural  
366             1.0         Rural  

[367 rows x 12 columns]

In [23]:
df.shape

(367, 12)

In [24]:
df_copy=df.copy()
df_copy

Loan_ID Gender Married Dependents     Education Self_Employed  \
0    LP001015   Male     Yes          0      Graduate            No   
1    LP001022   Male     Yes          1      Graduate            No   
2    LP001031   Male     Yes          2      Graduate            No   
3    LP001035   Male     Yes          2      Graduate            No   
4    LP001051   Male      No          0  Not Graduate            No   
..        ...    ...     ...        ...           ...           ...   
362  LP002971   Male     Yes         3+  Not Graduate           Yes   
363  LP002975   Male     Yes          0      Graduate            No   
364  LP002980   Male      No          0      Graduate            No   
365  LP002986   Male     Yes          0      Graduate            No   
366  LP002989   Male      No          0      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5720                  0       110.0             360.0   
1               3076               1500       126.0             360.0   
2               5000               1800       208.0             360.0   
3               2340               2546       100.0             360.0   
4               3276                  0        78.0             360.0   
..               ...                ...         ...               ...   
362             4009               1777       113.0             360.0   
363             4158                709       115.0             360.0   
364             3250               1993       126.0             360.0   
365             5000               2393       158.0             360.0   
366             9200                  0        98.0             180.0   

     Credit_History Property_Area  
0               1.0         Urban  
1               1.0         Urban  
2               1.0         Urban  
3               NaN         Urban  
4               1.0         Urban  
..              ...           ...  
362             1.0         Urban  
363             1.0         Urban  
364             NaN     Semiurban  
365             1.0         Rural  
366             1.0         Rural  

[367 rows x 12 columns]

In [25]:
df.isnull().sum()

Loan_ID               0
Gender               11
Married               0
Dependents           10
Education             0
Self_Employed        23
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount            5
Loan_Amount_Term      6
Credit_History       29
Property_Area         0
dtype: int64

In [26]:
df['Gender'].fillna(df['Gender'].mode()[0], inplace=True)
df['Dependents'].fillna(df['Dependents'].mode()[0], inplace=True)
df['Self_Employed'].fillna(df['Self_Employed'].mode()[0], inplace=True)
df['Credit_History'].fillna(df['Credit_History'].mode()[0], inplace=True)


# Test Numerical Variables Missing Values
df['Loan_Amount_Term'].fillna(df['Loan_Amount_Term'].mode()[0], inplace=True)
df['LoanAmount'].fillna(df['LoanAmount'].median(), inplace=True)
df['Gender'].fillna(df['Gender'].mode()[0], inplace=True)
df['Dependents'].fillna(df['Dependents'].mode()[0], inplace=True)
df['Self_Employed'].fillna(df['Self_Employed'].mode()[0], inplace=True)
df['Credit_History'].fillna(df['Credit_History'].mode()[0], inplace=True)


# Test Numerical Variables Missing Values
df['Loan_Amount_Term'].fillna(df['Loan_Amount_Term'].mode()[0], inplace=True)
df['LoanAmount'].fillna(df['LoanAmount'].median(), inplace=True)


In [27]:
df.isnull().sum()

Loan_ID              0
Gender               0
Married              0
Dependents           0
Education            0
Self_Employed        0
ApplicantIncome      0
CoapplicantIncome    0
LoanAmount           0
Loan_Amount_Term     0
Credit_History       0
Property_Area        0
dtype: int64

In [28]:
from sklearn.preprocessing import LabelEncoder
l=["Gender","Married","Education","Self_Employed","Property_Area"]

for i in l:
    le = LabelEncoder()
    le.fit(df[i])
    le.classes_
    
    df[i]=le.transform(df[i])

In [29]:
df

Loan_ID  Gender  Married Dependents  Education  Self_Employed  \
0    LP001015       1        1          0          0              0   
1    LP001022       1        1          1          0              0   
2    LP001031       1        1          2          0              0   
3    LP001035       1        1          2          0              0   
4    LP001051       1        0          0          1              0   
..        ...     ...      ...        ...        ...            ...   
362  LP002971       1        1         3+          1              1   
363  LP002975       1        1          0          0              0   
364  LP002980       1        0          0          0              0   
365  LP002986       1        1          0          0              0   
366  LP002989       1        0          0          0              1   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0               5720                  0       110.0             360.0   
1               3076               1500       126.0             360.0   
2               5000               1800       208.0             360.0   
3               2340               2546       100.0             360.0   
4               3276                  0        78.0             360.0   
..               ...                ...         ...               ...   
362             4009               1777       113.0             360.0   
363             4158                709       115.0             360.0   
364             3250               1993       126.0             360.0   
365             5000               2393       158.0             360.0   
366             9200                  0        98.0             180.0   

     Credit_History  Property_Area  
0               1.0              2  
1               1.0              2  
2               1.0              2  
3               1.0              2  
4               1.0              2  
..              ...            ...  
362             1.0              2  
363             1.0              2  
364             1.0              1  
365             1.0              0  
366             1.0              0  

[367 rows x 12 columns]

In [30]:
df=df.drop(["Loan_ID"],axis=1)
df

Gender  Married Dependents  Education  Self_Employed  ApplicantIncome  \
0         1        1          0          0              0             5720   
1         1        1          1          0              0             3076   
2         1        1          2          0              0             5000   
3         1        1          2          0              0             2340   
4         1        0          0          1              0             3276   
..      ...      ...        ...        ...            ...              ...   
362       1        1         3+          1              1             4009   
363       1        1          0          0              0             4158   
364       1        0          0          0              0             3250   
365       1        1          0          0              0             5000   
366       1        0          0          0              1             9200   

     CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0                    0       110.0             360.0             1.0   
1                 1500       126.0             360.0             1.0   
2                 1800       208.0             360.0             1.0   
3                 2546       100.0             360.0             1.0   
4                    0        78.0             360.0             1.0   
..                 ...         ...               ...             ...   
362               1777       113.0             360.0             1.0   
363                709       115.0             360.0             1.0   
364               1993       126.0             360.0             1.0   
365               2393       158.0             360.0             1.0   
366                  0        98.0             180.0             1.0   

     Property_Area  
0                2  
1                2  
2                2  
3                2  
4                2  
..             ...  
362              2  
363              2  
364              1  
365              0  
366              0  

[367 rows x 11 columns]

In [31]:
df[["Dependents"]]=df["Dependents"].replace(to_replace={'3+':3})
df.head()

Gender  Married Dependents  Education  Self_Employed  ApplicantIncome  \
0       1        1          0          0              0             5720   
1       1        1          1          0              0             3076   
2       1        1          2          0              0             5000   
3       1        1          2          0              0             2340   
4       1        0          0          1              0             3276   

   CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0                  0       110.0             360.0             1.0   
1               1500       126.0             360.0             1.0   
2               1800       208.0             360.0             1.0   
3               2546       100.0             360.0             1.0   
4                  0        78.0             360.0             1.0   

   Property_Area  
0              2  
1              2  
2              2  
3              2  
4              2

In [32]:
# Predict response fot test set
pred = knn.predict(df)
print(pred)
print(pred.shape)

[1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1
 0 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 0 1 1 1 0 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 0 0 0 0 1 0 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1
 1 1 0 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 1 0 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 0 1 1 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0
 0 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 1
 1 1 1 1 1 1 1 0 1 1 1 1 1 0 1 0 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 0 1 1 1 0 1 0 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
(367,)


In [34]:
submission = pd.DataFrame({"Loan_ID":df_copy["Loan_ID"], "Loan_Status": pred })


In [35]:
submission['Loan_Status'].replace(0, 'N',inplace=True) 
submission['Loan_Status'].replace(1, 'Y',inplace=True)

In [36]:
submission.to_csv('submission_loan_knn.csv', index=False)